### PR TITLE
[MRG] small ICA bug fixes for #5967 - 2nd try

### PR DESCRIPTION
**Pull request by jeythekey**
Continues/replaces #5968

### CODE REVIEW

File: mne/preprocessing/ica.py [hunk 1]
```diff
@@ -899,7 +909,7 @@ def _export_info(self, info, container, add_channels):
             # re-append additionally picked ch_info
             ch_info += [k for k in container.info['chs'] if k['ch_name'] in
                         add_channels]
-        info['bads'] = [ch_names[k] for k in self.exclude]
+        info['bads'] = [ch_names[k] for k in list(self.exclude)]
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: Why is this necessary?

File: mne/preprocessing/ica.py [hunk 2]
```diff
@@ -1367,7 +1377,8 @@ def _check_exclude(self, exclude):
         if exclude is None:
             return list(set(self.exclude))
         else:
-            return list(set(self.exclude + exclude))
+            return list(set(list(self.exclude) + list(exclude)))
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: or `return list(set(self.exclude).union(set(exclude)))`?

File: mne/preprocessing/ica.py [hunk 3]
```diff
@@ -1451,10 +1462,9 @@ def _apply_evoked(self, evoked, include, exclude, n_pca_components):
 
     def _pick_sources(self, data, include, exclude):
         """Aux function."""
-        if exclude is None:
-            exclude = self.exclude
-        else:
-            exclude = list(set(self.exclude + list(exclude)))
+        # We assume
+        # exclude = self._check_exclude(exclude)
+        # has already been called.
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: Why 3 lines?

File: mne/preprocessing/ica.py [hunk 4]
```diff
@@ -2171,10 +2181,13 @@ def _detect_artifacts(ica, raw, start_find, stop_find, ecg_ch, ecg_score_func,
         if isinstance(node.criterion, float):
             found = list(np.where(np.abs(scores) > node.criterion)[0])
         else:
-            found = list(np.atleast_1d(abs(scores).argsort()[node.criterion]))
+            found = list(np.atleast_1d(
+                    (-abs(scores)).argsort()[node.criterion]))
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: use `abs(scores)[::-1].argsort()` I think

File: mne/preprocessing/ica.py [hunk 5]
```diff
@@ -2171,10 +2181,13 @@ def _detect_artifacts(ica, raw, start_find, stop_find, ecg_ch, ecg_score_func,
...
+            # sort in descending order
 
         case = (len(found), _pl(found), node.name)
         logger.info('    found %s artifact%s by %s' % case)
+        ica.exclude = list(ica.exclude)
         ica.exclude += found
```
Comment (jona-sassenhagen) [inline mne/preprocessing/ica.py]: `ica.exclude = list(ica.exclude) + found`

File: mne/preprocessing/ica.py (hunk 1 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: See here: #5967

Comment (jona-sassenhagen) [reply mne/preprocessing/ica.py]: I don't see it.

File: mne/preprocessing/ica.py (hunk 2 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: Right. That would save 2 empty spaces ;)

File: mne/preprocessing/ica.py (hunk 3 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: Thanks

File: mne/preprocessing/ica.py (hunk 2 above)
Comment (jona-sassenhagen) [reply mne/preprocessing/ica.py]: No, it's more directly what you want to do right?

File: mne/preprocessing/ica.py (hunk 4 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: (It ate my reply):
You mean `abs(scores).argsort()[::-1]`?
That would also work and theoretically produce less overhead, but is also less accurate with the order of NANs and same values (probably both not relevant here). See here: https://stackoverflow.com/questions/16486252/is-it-possible-to-use-argsort-in-descending-order

File: mne/preprocessing/ica.py (hunk 5 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: Yes, obviously.

File: mne/preprocessing/ica.py (hunk 4 above)
Comment (jona-sassenhagen) [reply mne/preprocessing/ica.py]: exclude is an iterable of ints as of by this PR, so no NaNs.

The stackoverflow discussion also suggests [::-1] is better right?

File: mne/preprocessing/ica.py (hunk 2 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: The set() is around both lists, so it should have the same effect. But I have already changed it to your suggestion, because it looks nicer.

File: mne/preprocessing/ica.py (hunk 1 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: It was again eating my (2nd) comment ... Yes, you are right, not necessary here.

File: mne/preprocessing/ica.py (hunk 4 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: Well, it's sorting the scores. Can we always be sure no NaNs will occur there? Probably yes, but why bother?
The accepted answer says: "if a stable sort implementation is requested to argsort, [...] then the first strategy [== (-abs(...))] will preserve the sorting stability, but the second strategy will break stability (i.e. the positions of equal items will get reversed)."
Do you still want me to change it?

Comment (jona-sassenhagen) [reply mne/preprocessing/ica.py]: It's just a stylistic comment.

Can we be sure no NaNs will occur? No. But we shouldn't guarantee sensible sorting when the docstring says ints and people supply something not an int either.

Comment (jeythekey) [reply mne/preprocessing/ica.py]: Maybe I misunderstand something, but I thought the comment below the answer would imply that if any ICA-component yields a score of NaN by the method called in ica.score_sources, that component would be treated as the one with the highest score in the [::-1] version. Since we cannot be sure about the scoring func (it can also be a user provided one), I feel it's "safer" as it is.

Comment (jona-sassenhagen) [reply mne/preprocessing/ica.py]: No, it's my fault - I had somehow thought we're dealing with picks here, not scores. You're right.

**Comment by jona-sassenhagen** [comment]
Did you also check why the unit tests were falling?

**Comment by jona-sassenhagen** [comment]
Very good, you only have a pep failure:

https://travis-ci.org/mne-tools/mne-python/jobs/525415877

>mne/preprocessing/ica.py:2184:21: E126 continuation line over-indented for hanging indent

Fix that and set the title from FIX to MRG please.

File: doc/whats_new.rst [hunk 6]
```diff
@@ -205,6 +205,12 @@ Bug
 
 - Fix :meth:`mne.io.Raw.interpolate_bads` for interpolating CTF MEG channels when reference sensors are present by `jeythekey`_
 
+- Fix :meth:`mne.preprocessing.ICA.score_sources` to use the ``sfreq`` of the raw data to filter rather than the ``sfreq`` when the ICA was fit in :func:`mne.preprocessing.ICA._band_pass_filter` by `jeythekey`_
```
Comment (agramfort) [inline doc/whats_new.rst]: we don't document in what's new the private functions. These are not visible to end users. Only document changes to public API. thx

**Comment by jona-sassenhagen** [comment]
```
/home/circleci/project/doc/whats_new.rst:210: WARNING: py:attr reference target not found: mne.preprocessing.ICA.exclude
/home/circleci/project/doc/whats_new.rst:212: WARNING: py:func reference target not found: mne.preprocessing.ICA.run_ica
```

**Comment by jeythekey** [comment]
hm... and why? How to fix this? Is it important? run_ica is not part of the ICA class but in the same file. So how to link to this? Why it does not find the attribute I cannot see.

**Comment by larsoner** [comment]
> hm... and why? How to fix this?

You need to link to things where they are documented (implicitly or explicitly) in `python_reference.rst`, not where they end up in code. So the first one is `mne.preprocessing.run_ica`. The second one we actually can't link to without making it a `@property` rather than just variable, so for now instead of linking just use double backticks on it.

> Is it important?

The way it is now, we will have broken links in our `whats_new.html` on the website.

**Comment by jeythekey** [comment]
Thanks for the explanation. The first one was obvious, sorry ...

**Comment by larsoner** [comment]
@mmagnuski if you use these functions can you take a look?

File: mne/preprocessing/ica.py [hunk 7]
```diff
@@ -249,11 +249,14 @@ class ICA(ContainsMixin):
...
-              especially during preprocessing [5]_.
+              especially during preprocessing [5]_. (But see also [6]_ for a
+              possibly confounding effect of the different whitening/sphering
+              methods used in this paper (ZCA vs. PCA).)
               On the other hand, for rank-deficient data such as EEG data after
-              average reference or interpolation, it is recommended to reduce
-              the dimensionality by 1 for optimal ICA performance
-              (see the `EEGLAB wiki <eeglab_wiki_>`_).
+              average reference or interpolation, it is  recommended to reduce
+              the dimensionality (by 1 for  average reference and 1 for each
```
Comment (mmagnuski) [inline mne/preprocessing/ica.py]: two spaces after `for`. Good you added that, I though it Clemens included that earlier.

File: mne/preprocessing/ica.py [hunk 8]
```diff
@@ -1367,7 +1377,8 @@ def _check_exclude(self, exclude):
         if exclude is None:
             return list(set(self.exclude))
         else:
-            return list(set(self.exclude + exclude))
+            return list(set(self.exclude).union(set(exclude)))
+            # Allow both self.exclude and exclude to be array-like.
```
Comment (mmagnuski) [inline mne/preprocessing/ica.py]: It would be better if the comment preceded the code line it refers to, currenlty it is the last line in a method which is a bit weird. :)

File: mne/preprocessing/ica.py [hunk 9]
```diff
@@ -977,7 +987,7 @@ def score_sources(self, inst, target=None, score_func='pearsonr',
                                  'number of time slices.')
             # auto target selection
             if isinstance(inst, BaseRaw):
-                sources, target = _band_pass_filter(self, sources, target,
+                sources, target = _band_pass_filter(inst, sources, target,
```
Comment (mmagnuski) [inline mne/preprocessing/ica.py]: Please add a comment that we pass inst, not self because sfreq can be different, it was not obvious for me at the first sight so a comment would be useful.

File: mne/preprocessing/ica.py [hunk 10]
```diff
@@ -1451,10 +1462,8 @@ def _apply_evoked(self, evoked, include, exclude, n_pca_components):
 
     def _pick_sources(self, data, include, exclude):
         """Aux function."""
-        if exclude is None:
-            exclude = self.exclude
-        else:
-            exclude = list(set(self.exclude + list(exclude)))
+        # We assume exclude = self._check_exclude(exclude)  has already been
```
Comment (mmagnuski) [inline mne/preprocessing/ica.py]: is this assumption safe? Or in other words: if self._check_exclude is always called by mne code before this one, we don't have to assume anything here. :) (this is a private function after all)

File: mne/preprocessing/ica.py [hunk 11]
```diff
@@ -1599,8 +1608,8 @@ def detect_artifacts(self, raw, start_find=None, stop_find=None,
                          ecg_ch=None, ecg_score_func='pearsonr',
                          ecg_criterion=0.1, eog_ch=None,
                          eog_score_func='pearsonr',
-                         eog_criterion=0.1, skew_criterion=-1,
-                         kurt_criterion=-1, var_criterion=0,
```
Comment (mmagnuski) [inline mne/preprocessing/ica.py]: Why do we change the default criteria here? This would change the behavior of `detect_artifacts` between versions without any warning.

File: mne/preprocessing/ica.py [hunk 12]
```diff
@@ -1643,7 +1652,7 @@ def detect_artifacts(self, raw, start_find=None, stop_find=None,
             the name of function supported by ICA or a custom function.
         ecg_criterion : float | int | list-like | slice
             The indices of the sorted skewness scores. If float, sources with
-            scores smaller than the criterion will be dropped. Else, the scores
+            scores greater than the criterion will be dropped. Else, the scores
```
Comment (mmagnuski) [inline mne/preprocessing/ica.py]: was this an error in the docs or do you change the behavior here?

File: mne/preprocessing/ica.py [hunk 13]
```diff
@@ -2171,11 +2180,13 @@ def _detect_artifacts(ica, raw, start_find, stop_find, ecg_ch, ecg_score_func,
         if isinstance(node.criterion, float):
             found = list(np.where(np.abs(scores) > node.criterion)[0])
         else:
-            found = list(np.atleast_1d(abs(scores).argsort()[node.criterion]))
+            found = list(np.atleast_1d((-np.abs(scores)).argsort()
+                         [node.criterion]))
+            # sort in descending order
```
Comment (mmagnuski) [inline mne/preprocessing/ica.py]: comment should precede the code

File: mne/preprocessing/tests/test_ica.py [hunk 14]
```diff
@@ -534,6 +534,16 @@ def f(x, y):
                              eog_ch=ch_name, skew_criterion=idx,
                              var_criterion=idx, kurt_criterion=idx)
 
+    # Make sure detect_artifacts marks the right compos.
```
Comment (mmagnuski) [inline mne/preprocessing/tests/test_ica.py]: `compos -> components`

File: mne/preprocessing/ica.py (hunk 7 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: ... and before recommended ;)

File: mne/preprocessing/ica.py (hunk 8 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: ok

File: mne/preprocessing/ica.py (hunk 9 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: ok

File: mne/preprocessing/ica.py (hunk 10 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: I turned the comment into the required line of code and removed that in the calling functions, thereby removing duplications. OK?

File: mne/preprocessing/ica.py (hunk 11 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: Well, my reasoning was that this way we at least actually _keep_ the effect of the default values, because before this PR, [0] would index the lowest score and [-1] the highest, and know it is inverted (i.e. now it is as it is (and used to be) documented).

But you are right, it would be probably good to introduce a warning. Could you point me to a template for a warning about changing the behavior of a function? After all the bug is only in the incompatibility between documentation and code, so someone who knew the code and did not care about the docstring could be surprised ...

File: mne/preprocessing/ica.py (hunk 10 above)
Comment (mmagnuski) [reply mne/preprocessing/ica.py]: Yes, that should be ok, we'll say what tests have to say about that. :)

File: mne/preprocessing/ica.py (hunk 12 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: No, here I only changed the doc, since I felt what is actually done in the code (i.e. labelling scores exceeding a threshold) is intuitively what you would expect (in contrast to the above issue, where I found what was promised in the docstring is more intuitive that what was done in the code.)

File: mne/preprocessing/ica.py (hunk 13 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: ok

**Comment by jeythekey** [comment]
OK, I implemented your suggestion, thanks @mmagnuski. I also made the docstring a little more explicit.

To make the `detect_artifacts` method really useful, I think we need to (also) add the identified components to the `.labels_` attribute as it is done in `find_bads_eog/ecg` rather than just dumping everything into `.exclude`, so the reason for labelling is logged. (There is also already a nice inspection function which makes use of `.labels_`.) Also, it would probably make sense if `_detect_artifacts` would call those functions internally for eog and ecg artifacts, e.g. to make use of the default filtering etc. --> Ideas for a next PR :)

File: mne/preprocessing/ica.py [hunk 15]
```diff
@@ -987,6 +987,10 @@ def score_sources(self, inst, target=None, score_func='pearsonr',
                                  'number of time slices.')
             # auto target selection
             if isinstance(inst, BaseRaw):
+                # We need to pass inst, not self for defining sfreq, since the
+                # sfreq when fitting ica (--> self.info['sfreq']) can be
+                # different (usually lower) than the sfreq when scoring sources
+                # (e.g. using the original raw sfreq for this step):
```
Comment (mmagnuski) [inline mne/preprocessing/ica.py]: I would keep the comment short: something like `we pass inst, not self, because the sfreq of the data we apply ica to can be different`, all the other details are not necessary.

Comment (jeythekey) [reply mne/preprocessing/ica.py]: Yes, that I have to learn, to KISS ;)

File: mne/preprocessing/ica.py (hunk 12 above)
Comment (mmagnuski) [reply mne/preprocessing/ica.py]: sure, if the docstring had an error - ok.

**Comment by jeythekey** [comment]
OK, apparently, `SyntaxWarning` is the wrong class. What should it be instead? RuntimeWarning? I am not deprecating anything.

File: mne/preprocessing/ica.py (hunk 11 above)
Comment (larsoner) [reply mne/preprocessing/ica.py]: The least surprise from a user perspective would be to keep the old behavior and old defaults, and update the docstring to say what this does.

If our old defaults made no sense, we can have a deprecation cycle where we set them to `None` for a cycle in the signature, and in 0.18 `None` does what was done in 0.17 (with a warning that not setting it explicitly will change in 0.19) and then in 0.19 we change the default to something sane.

Comment (mmagnuski) [reply mne/preprocessing/ica.py]: I agree with @larsoner, BTW @jeythekey - you should better ask Eric about how to implement deprecations, I never did this so I would have to look around in our codebase, but Eric knows all this off the top of his head.

Comment (jeythekey) [reply mne/preprocessing/ica.py]: So, in principle we would revert this PR and replace it with a warning, that in the next release the function will do what it used to say it would do in the docstring?
Is this really necessary in this case, where the docstring clearly states "Note. This is still experimental and will most likely change over the next releases."?
As it is now, at least the defaults will do the same as before (they were only changed, because of the bug-fix in the code), only when the criteria args used to be set contrary to what would be sensible given the docstring by a user, the behavior will change. So, this should only negatively affect those who dug through the code before, identified the bug and decided to silently adapt their usage instead of reporting the bug.

So, please, one last word by @larsoner: Should I really revert the fix for sorting the compos in descending order and replace the fix by a warning about a future fix?

Comment (mmagnuski) [reply mne/preprocessing/ica.py]: could you explain the sorting bug you mention? The whats new entry says "Fix descending sorting of scores for integer" - why do we want to change the sorting, why is descending not ok?

Comment (jeythekey) [reply mne/preprocessing/ica.py]: The doc said "descending", but the code did "ascending". So `skew_criterion=0` returned the score with the smallest absolute skewness, contrary to what was further explained in the doc by "E.g. range(2) would return the two sources with the highest score." I find it intuitive as the docstring says, so I change the code to do just that. But of course I am open to change it to your liking! If you think we should stick to ascending sorting I would just change the doc.

Comment (mmagnuski) [reply mne/preprocessing/ica.py]: No, descending is good, it should be descending, so this is a bug fix. Then:
* change the what's new to state that ascending was changed into descending. I read the "Fix descending sorting" as if it was descending and was changed.
* I am ok with changing the criteria now, otherwise it would be nonsense. I somehow didn't before get that these "criteria" are indices. What do you think @larsoner? In case someone actually uses these we could set these to None and issue a warning.

Comment (larsoner) [reply mne/preprocessing/ica.py]: > In case someone actually uses these we could set these to None and issue a warning.

Even in this case if someone had previously passed values, now the behavior will be changed. I guess since the behavior is wrong and this is a bugfix, we can change the behavior like you say and make it clear in the BUG section that this has been fixed.

@agramfort are you okay with this?

Comment (jeythekey) [reply mne/preprocessing/ica.py]: > Even in this case if someone had previously passed values ...

Actually, _only_ in this case ... (because the default values were inverted to retain the old behavior - and they make sense now, I guess, as they did before)

And what do you think about [this warning](https://github.com/mne-tools/mne-python/pull/6246/commits/1aacc1f59c0acfec0f3f7b5e93e53eb8695f82bf#diff-5a816b651c2f6b26c8899e97bfe6549aR2153) (at the end) I tried to implement (with the wrong warning class, apparently)?

File: mne/preprocessing/ica.py [hunk 16]
```diff
@@ -214,11 +214,11 @@ class ICA(ContainsMixin):
         If fit, the whitened matrix to go from PCA space to ICA space.
         Used, in combination with the `pca_components_`, by the methods
         :meth:`ICA.get_sources` and :meth:`ICA.apply` to unmix the observed data.
-    exclude : list
-        List of sources indices to exclude when re-mixing the data in the
-        :meth:`ICA.apply` method, i.e. artifactual ICA components.
+    exclude : array_like of int
```
Comment (agramfort) [inline mne/preprocessing/ica.py]: array_like -> array-like

File: mne/preprocessing/ica.py [hunk 17]
```diff
@@ -1367,12 +1379,12 @@ def _check_exclude(self, exclude):
...
             return list(set(self.exclude))
         else:
-            return list(set(self.exclude + exclude))
+            # Allow both self.exclude and exclude to be array-like:
+            return list(set(self.exclude).union(set(exclude)))
 
     def _apply_raw(self, raw, include, exclude, n_pca_components, start, stop):
         """Aux method."""
         _check_preload(raw, "ica.apply")
-        exclude = self._check_exclude(exclude)
```
Comment (agramfort) [inline mne/preprocessing/ica.py]: why removing all these calls to _check_exclude?

Comment (jeythekey) [reply mne/preprocessing/ica.py]: ... because this function is called in _pick_sources, which is called by all the apply_... funcs. Since _pick_sources does the exclude-relevant work I found it more consistent and future-proof to call _check_exclude there.

File: mne/preprocessing/ica.py (hunk 11 above)
Comment (jeythekey) [reply mne/preprocessing/ica.py]: @larsoner and @mmagnuski should I then just remove the warning? I changed the bug section as you suggested. Because then we should be good to go, or is there anything else, @agramfort ?

Comment (larsoner) [reply mne/preprocessing/ica.py]: Yes let's remove the warning if it's a bug. Then wait for @agramfort to comment next week to make sure it's reasonable

**Comment by agramfort** [comment]
@mmagnuski or @larsoner merge if you're happy. Travis error is unrelated

**Comment by larsoner** [comment]
I'll let you take a last look and merge if you're happy @mmagnuski

**Comment by mmagnuski** [comment]
I'm happy, thanks @jeythekey!

**Comment by GuillaumeFavelier** [comment]
@jeythekey we don't have your name to update the mailing for the release. Can you provide us your full name and desired email to ament that? thx.

**Comment by jeythekey** [comment]
@GuillaumeFavelier : Johannes Kasper, jeythekey@tutanota.com
